### ISSUE TITLE
Provide scaler for Elasticsearch

### ISSUE THREAD

**Issue by tomkerkhove** [kedacore/keda#1756]
### Proposal

Provide scaler for Elasticsearch to scale on your metrics:

https://www.elastic.co/infrastructure-monitoring

### Scaler Source

Elasticsearch

### Scaling Mechanics

Scale based on query outcome

### Authentication Source

TBD

**Comment by kozdincer** [comment]
I think there are a lot of query types in elasticsearch to configure. Query structures also greatly depending on the document settings.[0]
I think it is very complicated to write a scaler specific to ES.

It can be worked on a generic http scaler with URL setting, CRUD requests (GET queries and POST data), the response handling and necessary auth mechanisms (basic auth, oauth etc.) can be implemented.

Thus, I think that metrics in many data sources with a REST interface or custom metrics can be used.

What you think?

[0] https://www.tutorialspoint.com/elasticsearch/elasticsearch_query_dsl.htm

**Comment by kozdincer** [comment]
I can prepare a more detailed proposal if you are interested. And there we can discuss there.

**Comment by tomkerkhove** [comment]
@melisatanrverdi is working on this

**Comment by tomkerkhove** [comment]
@kozdincer I see you've posted some input but have removed it; was it on purpose?

We have @melisatanrverdi working on it but feel free to share thoughts

**Comment by kozdincer** [comment]
There are too many query types for ES[0]. A wide variety of structured data is sent to the GET request. I don't know how we can make this into the YAML config.

Still, it seemed more appropriate to provide support for SQL[1] and EQL[2]. The only problem here is that EQL and SQL supports are "x-pack"[2]. What are your opinions? @tomkerkhove 
@melisatanrverdi

[0] https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html
[1] https://www.elastic.co/guide/en/elasticsearch/reference/current/xpack-sql.html
[2] https://www.elastic.co/guide/en/elasticsearch/reference/master/eql.html
[3] https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-xpack.html

**Comment by melisatanrverdi** [comment]
Hi, I think the same as @kozdincer . There are so many query types[0] and also they are quite changeable. While one can takes 2 variables, the other can takes 10. And their position can also change. That means the POST is also different for each.
It seems like the real struggle is we have to handle lots of YAML file besides scaler itself[1]. 
So we thought two things:
1. We can provide POST support to Metrics API[2].
2. We can write Elasticsearch scaler with EQL and SQL support.
But like @kozdincer  said the thing for second one is EQL and SQL supports are "x-pack"[3]. We are not sure this is a problem or not.
We would like to hear your suggestions. 

[0] https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html
[1] https://github.com/olivere/elastic
[2] https://keda.sh/docs/2.2/scalers/metrics-api/
[3] https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-xpack.html

**Comment by tomkerkhove** [comment]
What do you think @zroubalik?

**Comment by tomkerkhove** [comment]
I'm reluctant for just using the MEtrics API to be honest. If we cannot provide a straight forward scaler spec, then I think it's best to just abandon it which would be a pity.

**Comment by zroubalik** [comment]
> I'm reluctant for just using the MEtrics API to be honest. If we cannot provide a straight forward scaler spec, then I think it's best to just abandon it which would be a pity.

Yeah, I agree we. We should get back to this scaler only if there's increased demand.

@melisatanrverdi great job on the investigation!

**Comment by burkel24** [comment]
Just chiming in to say that I would be interested using an ES scaler, though agree with above points re: complexity of universality.

Planning to implement an external scaler instead tailored to my usecase. Please feel free to reach out if you guys decide to pick this up and want context about my usecase.
